lib/Basics/ReadWriteLock.cpp

```diff
@@ -27,79 +27,103 @@ using namespace arangodb::basics;
 
 /// @brief locks for writing
 void ReadWriteLock::writeLock() {
-  std::unique_lock<std::mutex> guard(_mut);
-  if (_state == 0) {
-    _state = -1;
+  if (tryWriteLock())
     return;
+
+  // the lock is either hold by another writer or we have active readers
+  // -> announce that we want to write
+  _state.fetch_add(QUEUED_WRITER_INC, std::memory_order_relaxed);
+
+  std::unique_lock<std::mutex> guard(_writer_mutex);
+  while (true)
+  {
+    auto state = _state.load(std::memory_order_relaxed);
+    while ((state & ~QUEUED_WRITER_MASK) == 0)
+    {
+      // try to acquire lock and perform queued writer decrement in one step
+      if (_state.compare_exchange_strong(state, (state - QUEUED_WRITER_INC) | WRITE_LOCK, std::memory_order_acquire))
+        return;
+    }
+    _writers_bell.wait(guard);
   }
-  do {
-    _wantWrite = true;
-    _bell.wait(guard);
-  } while (_state != 0);
-  _state = -1;
-  _wantWrite = false;
 }
 
 /// @brief locks for writing, but only tries
 bool ReadWriteLock::tryWriteLock() {
-  std::unique_lock<std::mutex> guard(_mut);
-  if (_state == 0) {
-    _state = -1;
-    return true;
+  auto state = _state.load(std::memory_order_relaxed);
+  // try to acquire write lock as long as no readers or writers are active,
+  // we might "overtake" other queued writers though.
+  while ((state & ~QUEUED_WRITER_MASK) == 0)
+  {
+    if (_state.compare_exchange_weak(state, state | WRITE_LOCK, std::memory_order_acquire)) {
+      return true; // we successfully acquired the write lock!
+    }
   }
   return false;
 }
 
 /// @brief locks for reading
 void ReadWriteLock::readLock() {
-  std::unique_lock<std::mutex> guard(_mut);
-  if (!_wantWrite && _state >= 0) {
-    _state += 1;
+  if (tryReadLock())
     return;
+
+  std::unique_lock<std::mutex> guard(_reader_mutex);
+  while (true)
+  {
+    if (tryReadLock())
+      return;
+
+    _readers_bell.wait(guard);
   }
-  while (true) {
-    while (_wantWrite || _state < 0) {
-      _bell.wait(guard);
-    }
-    if (!_wantWrite) {
-      break;
-    }
-  }
-  _state += 1;
 }
 
 /// @brief locks for reading, tries only
 bool ReadWriteLock::tryReadLock() {
-  std::unique_lock<std::mutex> guard(_mut);
-  if (!_wantWrite && _state >= 0) {
-    _state += 1;
-    return true;
+  auto state = _state.load(std::memory_order_relaxed);
+  // try to acquire write lock as long as no writers are active or queued
+  while ((state & ~READER_MASK) == 0) {
+    if (_state.compare_exchange_weak(state, state + READER_INC, std::memory_order_acquire)) {
```

Comment:
why use the `_weak` version here ?

Reply:
ah nevermind its supposed to be more efficient in loops

Reply:
exactly! compare_exchange_weak can be more efficient on some architectures, but might fail spuriously. But when it is used in a loop this doesn't matter as we simply retry in the next iteration.  (not that this makes any difference on Intel based architectures since everything translates to a sequentially consistent cmpxchg instruction anyway 🙂 )